WreckEngine/Text.cpp

```diff
@@ -1,198 +1,549 @@
-		auto a = glyph.bearing.y * scale
-		   , b = glyph.size.y * scale - a;
-		if (a > above) above = a;
-		if (b > below) below = b;
-	};
-	
-	auto len = text.length() - 1;
-	for (size_t i = 0; i < len; ++i) {
-		const auto glyph = font->glyphs.at(text[i]);	
-		x += (glyph.advance >> 6) * scale;
-		updateY(glyph);
-	}
-
-	const auto glyph = font->glyphs.at(text[len]);
-	x += (glyph.size.x + glyph.bearing.x) * scale;
-	updateY(glyph);
-
-	return vec2(x, above + below);
-}
-
-// TODO update this to use a method that doesn't rely on sharing a buffer that changes between draw calls
-void Text::Renderer::draw(Text::Instance& instance) {
-	shader.program.use();
-	shader.color.update(instance.color);
-	shader.cam.update(glm::ortho(0.f, (float)Window::width, 0.f, (float)Window::height));
-
-	vao.bind();
+Text::Instance::Instance() {
+    vao.create();
+    vao.bind();
+
+    buffer.create(GL_ARRAY_BUFFER, GL_DYNAMIC_DRAW);
+    buffer.bind();
+    buffer.data(0, nullptr);
+
+    GLattrarr attrs;
+    attrs.add<vec2>(2);
+    attrs.add<uint32_t>(1);
+    attrs.apply();
+
+    buffer.unbind();
+    vao.unbind();
+
+    auto& shader = Renderer::shader;
+    fullOffset = GLresource<vec2>(shader.offset);
+    scale = GLresource<float>(shader.scale);
+    scale.value = 1;
+    renderInfo.setShaders(shader.program, &shader.cam, &fullOffset, &scale);
+}
+
+void Text::Instance::queueForDraw()
+{
+    instances.push_back(this);
+}
+
+void Text::Instance::updateAlignment() {
+    if (!dirtyAlign)
+        return;
+
+    alignOffset = vec2(0);
+
+    if (vert != START || horiz != START) {
+
+        int lineCount; // Guaranteed to at least be 1
+        auto dims = getDims(text, font, scale.value, lineCount);
+        
+        switch (horiz)
+        {
+        case MIDDLE:
+            alignOffset.x = -dims.x * 0.5f;
+            break;
+        case END:
+            alignOffset.x = -dims.x;
+            break;
+        }
+
+        switch (vert)
+        {
+        case MIDDLE:
+            alignOffset.y = dims.y * 0.5f;
+            break;
+        case END:
+            alignOffset.y = -dims.y;
+            break;
+        }
+    }
+
+    dirtyAlign = false;
+}
+
+void Text::Instance::updateBuffer() {
+    if (!dirtyBuffer || !font) {
+        return;
+    }
+
+    // Get some helper variables
+    const uint32_t packedColor = Color::pack(color);
+    const float xSpace = font->spaceWidth;
+    const float ySpace = font->lineHeight;
+    float x = 0.0f;
+    float y = 0.0f;
+    uint32_t prevCP = 0;
+    std::vector<CharVertex> vertices;
+
+    // Now go through and create one quad per character
+    for (size_t i = 0, length = text.length(); i < length; ++i) {
+        uint32_t currCP = text[i];
+
+        // Apply the kerning between the previous and current character
+        x += getKerning(prevCP, currCP, font);
+        prevCP = currCP;
+
+        // Handle special characters
+        if (isspace(currCP)) {
```

Comment:
Because there are more whitespace characters than the ones I'm handling. And it's [here](http://en.cppreference.com/w/c/string/byte/isspace).